share/picongpu/tests/CurrentDeposition/lib/python/test/CurrentDeposition/grid_class.py

```diff
@@ -0,0 +1,175 @@
+        """Calculates the current deposition vector W component-wise
+        for z, y, x.
+        Returns the components of the vector in the x, y, z direction on the
+        grid. Takes the tuple of start and end coordinates (x, y, z) as well
+        as the zero grids for each coordinate.
+        """
+        func = self.func
+
+        for z in range(self.num_cells):
+            for y in range(self.num_cells):
+                for x in range(self.num_cells):
+                    # Calculate the si components for the W calculation
+
+                    # Old
+                    s1 = func(start_coord[0] - x)
+                    s2 = func(start_coord[1] - y)
+                    s3 = func(start_coord[2] - z)
+                    # new
+                    s4 = func(end_coord[0] - x)
+                    s5 = func(end_coord[1] - y)
+                    s6 = func(end_coord[2] - z)
+
+                    # calculate W and assign it to the grid node
+                    grid_x[z][y][x] = W(s1, s2, s3, s4, s5, s6)
+                    grid_y[z][y][x] = W(s2, s1, s3, s5, s4, s6)
+                    grid_z[z][y][x] = W(s3, s2, s1, s6, s5, s4)
+
+        return grid_x, grid_y, grid_z
+
+    def current_density_field(self, W_grid_x, W_grid_y, W_grid_z, start_coord,
+                              end_coord, Q, cell_depth, cell_width,
+                              cell_height, dt, unit_time, unit_charge,
+                              unit_length):
+        """Calculates the current density in the x, y, and z directions on the
+        grid using the current deposition vector(field) W_grid_x,y,z. Charge Q
+        is needed as input. dx, dy, dz, and dt are set to 1 by default.
+        """
+
+        c = const.c
+
+        # create null grid to write the current density
+        null_axis = np.zeros(self.num_cells, dtype=float)
+        j_grid_x, j_grid_y, j_grid_z = np.meshgrid(null_axis, null_axis,
+                                                   null_axis)
+
+        # determine how many grid points are needed to be included
+        # in calculation; aka half-width of assignment function
+        add_active = (self.order + 1) * 1/2
+
+        # determine start x
+        start_i = int(min(start_coord[0], end_coord[0]) - add_active)
+        start_j = int(min(start_coord[1], end_coord[1]) - add_active)
+        start_k = int(min(start_coord[2], end_coord[2]) - add_active)
+
+        # determine end x
+        end_i = int(max(start_coord[0], end_coord[0]) + add_active) + 1
+        end_j = int(max(start_coord[1], end_coord[1]) + add_active) + 1
+        end_k = int(max(start_coord[2], end_coord[2]) + add_active) + 1
+
+        combined_factors = (-Q * unit_charge) / (c * dt * unit_time)
+        x_factor = combined_factors * (cell_width * unit_length)
+
+        # calculate the norm of the assignment function
+        H = cell_width * cell_height * cell_depth * unit_length**3
+        norm = (1/H)**self.order

```

Comment:
Are you sure about the norm? Shouldn't it be (1/H)**(self.order+1)?

Reply:
Yes. It should be `(1/H)**self.order`.
As one can see in my Bachelors Thesis (p. 6)  or in R. W. Hockney, J. W. Eastwood, Computer Simulation Using Particles, (1st ed.), CRC Press, 1988 (p.143), the norm is "just" calculated for the cloud shape. The assignment function is just the convolution of the (normed) cloud shape with a top had function PI(x/H)